### ISSUE TITLE
gpt_oss inference activates *all* experts for every token

### ISSUE THREAD

**Issue by shivak** [huggingface/transformers#40073]
### System Info

N/A

### Who can help?

@SunMarc 

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

N/A

### Expected behavior

OpenAI [describes](https://openai.com/index/introducing-gpt-oss/) the gpt-oss models as sparsely activated:
> Each model is a Transformer which leverages mixture-of-experts (MoE[2]) to reduce the number of active parameters needed to process input. gpt-oss-120b activates 5.1B parameters per token, while gpt-oss-20b activates 3.6B. The models have 117b and 21b total parameters respectively. 

While reading this comment in the gpt-oss implementation, "all" unfortunately doesn't refer to just all of the top k:
https://github.com/huggingface/transformers/blob/99c747539e07b2e141cb5f13961b72108e9dc864/src/transformers/models/gpt_oss/modeling_gpt_oss.py#L82
The implementation indeed involves the weights of *all* experts (not just top k) for every token during inference: 
https://github.com/huggingface/transformers/blob/99c747539e07b2e141cb5f13961b72108e9dc864/src/transformers/models/gpt_oss/modeling_gpt_oss.py#L119
(Recall that gpt-oss-120b's [configuration](https://huggingface.co/openai/gpt-oss-120b/blob/main/config.json) has `num_experts_per_tok=4` and `num_local_experts=128`)
https://github.com/huggingface/transformers/blob/99c747539e07b2e141cb5f13961b72108e9dc864/src/transformers/models/gpt_oss/modeling_gpt_oss.py#L67-L71
The training branch does route only to the top k experts, as in Mixtral's sparse MoE implementation (which was carried over to Qwen2 MoE and Qwen3 MoE).
https://github.com/huggingface/transformers/blob/99c747539e07b2e141cb5f13961b72108e9dc864/src/transformers/models/gpt_oss/modeling_gpt_oss.py#L97-L104

**Comment by qrdlgit** [comment]
Nice catch, are you using kernels (if so, note the execution path is different for routing)?  Is this just source analysis or did you profile the behavior?

**Comment by ArthurZucker** [comment]
That is expected, it's the most efficient way to compute this. If you want a more memory efficient you can use `use_kernels=True` when loading the model, it will use https://huggingface.co/kernels-community/megablocks/tree/main/torch-ext/megablocks !

**Comment by shivak** [comment]
True, in some circumstances, the bandwidth/compute/latency tradeoff of this implementation is favorable. But maybe not during prefill on platforms with lots of memory but not much bandwidth or TOPS (where MoE is supposed to shine). Also, gpt-oss-120b has 4/128 active experts, so we are talking about a lot of wasted work

I agree the Python perf characteristics are not ideal either way. But the rest of the MoE models use a for loop - I think some uniformity here is warranted. A unified solution to the sequential for loop can/should be applied to all of them

**Comment by ArthurZucker** [comment]
You can also just set model.train() which will not add gradients or anything but will use the training path

**Comment by ikergarcia1996** [comment]
Similarly, for sliding attention, the eager implementation computes the full attention matrix and then zeroes-out the values outside the sliding window. I assume this is for the same reason that all experts are computed (probably hard to implement a true sliding window efficiently in torch), although I don’t think many people actually use the eager implementation.

**Comment by ArthurZucker** [comment]
100% 😉 We have a few variations that do for loops for the attention but its not really more efficient.
